tests/e2e.test.ts

```diff
@@ -0,0 +1,268 @@
+ * This middleware catches Ganache's custom revert error and returns it as response data instead.
+ */
+class RevertNormalisingMiddleware extends ethers.providers.BaseProvider {
+  readonly parent: ethers.providers.BaseProvider;
+
+  constructor(provider: ethers.providers.BaseProvider) {
+    super(provider.getNetwork());
+    this.parent = provider;
+  }
+
+  getSigner(addressOrIndex?: string | number): JsonRpcSigner {
+    return (this.parent as Web3Provider).getSigner(addressOrIndex);
+  }
+
+  async perform(method: string, params: any): Promise<any> {
+    switch (method) {
+      case 'call':
+        try {
+          return await this.parent.perform(method, params);
+        } catch (e) {
+          if (isOffchainLookupError(e)) {
+            return e.error.data;
+          }
+          throw e;
+        }
+      default:
+        const result = await this.parent.perform(method, params);
+        return result;
+    }
+  }
+
+  detectNetwork(): Promise<ethers.providers.Network> {
+    return this.parent.detectNetwork();
+  }
+}
+
+describe('End to end test', () => {
+  const server = makeApp(
+    dohQuery(process.env.DOH_GATEWAY_URL as string),
+    '/',
+    Server
+  );
+
+  async function fetcher(
+    _url: string,
+    json?: string
+  ) {
+    const { sender: to, data } = JSON.parse(json as string);
+    const ret = await supertest(server).get(`/${to}/${data}.json`);
+    return ret;
+  }
+  const baseProvider = new ethers.providers.JsonRpcProvider(
+    'http://127.0.0.1:8545'
+  );
+  const signer = baseProvider.getSigner();
+  const proxyMiddleware = new RevertNormalisingMiddleware(baseProvider);
+  const mockProvider = new MockProvider(proxyMiddleware, fetcher);
+  let resolver: Contract, ownedResolver: Contract;
+
+  async function checkIfContractIsReachable(
+    address: string,
+    label: string
+  ): Promise<void> {
+    try {
+      const code = await baseProvider.getCode(address);
+      if (code !== '0x') console.log(`${label} contract is reachable.`);
+    } catch (error) {
+      console.log('getCode error', error);
+    }
+  }
+
+  beforeAll(async () => {
+    await checkIfContractIsReachable(ENS_ADDRESS, 'ENS');
+    await checkIfContractIsReachable(DNSSEC_IMPL, 'DNSSEC_IMPL');
+    await checkIfContractIsReachable(OWNED_RESOLVER, 'OWNED_RESOLVER');
+
+    resolver = (
+      await deploySolidity(
+        OffchainDNSResolver_abi,
+        signer,
+        ENS_ADDRESS,
+        DNSSEC_IMPL,
+        TEST_URL
+      )
+    ).connect(mockProvider);
+
+    ownedResolver = new ethers.Contract(
+      OWNED_RESOLVER,
+      Resolver_abi.abi,
+      signer
+    );
+  });
+
+  describe('resolve()', () => {
+    it('resolves calls to addr(bytes32)', async () => {
+      await ownedResolver['setAddr(bytes32,address)'](
+        ethers.utils.namehash(TEST_NAME),
+        TEST_ADDRESS
+      );
+
+      const callData = Resolver.encodeFunctionData('addr(bytes32)', [
```

Comment:
I'm a little confused as to what's going on here. Where is the call to the gateway happening? If it's via Ethers' built in CCIP-Read support, why do we then manually call `resolveCallback` ourselves?

Reply:
ccipRead is disabled by default in ethers (at least in v5.x). to be able to debug all the way down, I used supertest to make requests to the gateway, providing mockProvider, as we used to do in previous offchain resolver tests;

https://github.com/ensdomains/ccip-read-dns-gateway/blob/af575ee0b32b8f19ab632912aca7367cd9a7d057/tests/e2e.test.ts#L201

https://github.com/ensdomains/ccip-read-dns-gateway/blob/af575ee0b32b8f19ab632912aca7367cd9a7d057/tests/e2e.test.ts#L182-L186
 
https://github.com/ensdomains/ccip-read-dns-gateway/blob/af575ee0b32b8f19ab632912aca7367cd9a7d057/tests/e2e.test.ts#L188-L193

I tried adding another scenario being done with ethers.js only with ccip-read enable. but the following issue was a blocker, which I couldn't really understood, thought, having ccip gateway running locally might be the issue.

```bash
error encountered during CCIP fetch: "Internal server error: TypeError: Only absolute URLs are supported" (urls=["http://localhost:8000/query"], errorMessages=["Internal server error: TypeError: Only absolute URLs are supported"], code=SERVER_ERROR, version=providers/5.7.2)
```